WooCommerce/Classes/Tools/InfiniteScroll/PaginationTracker.swift

```diff
@@ -0,0 +1,152 @@
+import Yosemite
+
+/// Delegate of `PaginationTracker` that implements syncing per page number and size.
+protocol PaginationTrackerDelegate: class {
+    typealias SyncCompletion = (Result<Bool, Error>) -> Void
+
+    /// Syncs a page by the number and size, and returns an error or whether there might be the next page on completion.
+    ///
+    /// - Parameter reason: A value passed from `resync` or `syncFirstPage`. This can
+    ///                     be used to decide how to perform the sync.
+    /// - Parameter onCompletion: called on sync completion that returns either an error or a boolean that indicates
+    ///                           whether there might be the next page to sync.
+    func sync(pageNumber: Int, pageSize: Int, reason: String?, onCompletion: SyncCompletion?)
+}
+
+/// Keeps track of the pagination for API syncing to support infinite scroll.
+final class PaginationTracker {
+    /// Default Settings
+    private enum Defaults {
+        static let pageFirstIndex = Store.Default.firstPageNumber
+        static let pageSize = 25
+    }
+
+    /// First Page Index
+    private let pageFirstIndex: Int
+
+    /// Number of elements retrieved per request.
+    private let pageSize: Int
+
+    /// Syncing delegate
+    weak var delegate: PaginationTrackerDelegate?
+
+    /// Maps from a page number to the last synced date.
+    private var syncedDatesByPage = [Int: Date]()
```

Comment:
I'm thinking we should rename this to `pagesSynced` or something that's closer to `pagesBeingSynced` (below). I think they're closely related and it would help if they're similarly named.

Reply:
yup I renamed it to `pagesSynced` to be consistent with `pagesBeingSynced` and also because it has the same type as `pagesBeingSynced` now! updated in https://github.com/woocommerce/woocommerce-ios/pull/2750/commits/85d2cbc94229ee56a5487eb03b9996b1dfb943f8